[Ng Yong Chuan] Ip

yongchuann opened the pull request:
# Pika Bot:
>The solution to your (or my own) procrastination :clown_face: - Pikachu

Pika Bot is a lil bot that can do the following:
- [X] Track Task
- [x] Track deadlines
- [x]  Send you reminders
- [X]  ~~STOPS YOUR PROCRASTINATION by zapping you~~ :zany_face:

It also really easy to use!!! It is:
* text-based
* easy to learn and use
* fast as Flash
* ~~also SHOCKS you if you do something wrong~~

and most importantly.... ITS **FREEEEEEEEEEEEEE!**

So how do you get it? 
1. Visit his home [here](https://github.com/random689/ip) and invite him to your computer!
2. double-click it
3. add your task
4. let him do the rest!

`<Pika>`

Remember to put where you want the task list to be stored!
~~~
public class Main {
    public static void main(String[] args) {
       new Pika('data.txt');
    }
}
~~~

Commit:
Add Gradle support

yongchuann committed:
Implement a skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.

yongchuann committed:
Add the ability to mark tasks as done.
Also added in a task class to handle the tasks

yongchuann committed:
Add the ability to mark tasks as done.
Also added in a task class to handle the tasks

yongchuann committed:
Add support for tracking three types of tasks:
    ToDos: tasks without any date/time attached to it e.g., visit new theme park
    Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
    Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

 Also added the classes as well for each new task.
 Also updated the if else to switch-case
 Also removed the old base case and makes the bot give an re try response.

yongchuann committed:
Add support for tracking three types of tasks:
    ToDos: tasks without any date/time attached to it e.g., visit new theme park
    Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
    Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

 Also added the classes as well for each new task.
 Also updated the if else to switch-case
 Also removed the old base case and makes the bot give an re try response.

yongchuann committed:
Add support for tracking three types of tasks:
    ToDos: tasks without any date/time attached to it e.g., visit new theme park
    Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
    Events: tasks that start at a specific time and ends at a specific time e.g., team project meeting on 2/10/2019 2-4pm

 Also added the classes as well for each new task.
 Also updated the if else to switch-case
 Also removed the old base case and makes the bot give an retry response.

 added and used the textUItesting.

yongchuann committed:
Added DukeExceptions to catch the errors for:
1. Invalid command for "done", "todo", "deadline" and "event"
2. Catch invalid numbers (out of range) for "done"

Also updated to remove the extra space in each space (using trim)

yongchuann committed:
Added DukeExceptions to catch the errors for:
1. Invalid command for "done", "todo", "deadline" and "event"
2. Catch invalid numbers (out of range) for "done"

Also updated to remove the extra space in each space (using trim)

yongchuann committed:
Added DukeExceptions to catch the errors for:
1. Invalid command for "done", "todo", "deadline" and "event"
2. Catch invalid numbers (out of range) for "done"

Also updated to remove the extra space in each space (using trim)

yongchuann committed:
Added DukeExceptions to catch the errors for:
1. Invalid command for "done", "todo", "deadline" and "event"
2. Catch invalid numbers (out of range) for "done"

Also updated to remove the extra space in each space (using trim)

yongchuann committed:
Add support for deleting tasks from the list.

yongchuann committed:
Add the ability for the bot to save the list to harddisk

yongchuann committed:
Add the ability for the bot to recognise date time by adding a new TaskTime class to handle it

The way the time gated tasks handle the time should be the same for event and deadline task

yongchuann committed:
Merge branch 'branch-level-8'

yongchuann committed:
Add Ui, Command, TaskList, Parser, Storage class to make the code more OOP

Add the ability for the code to read an existing file and import the existing file's list to continue.

yongchuann committed:
Organise the class into duke package and their subpackages

yongchuann committed:
Add JUnit test to test the behaviour of the codes.

yongchuann committed:
Add JavaDoc comments
Tweak the code to comply with a coding standard

yongchuann committed:
Merge branch 'branch-A-CodingStandard' into branch-A-JavaDoc

yongchuann committed:
Merge branch 'branch-A-CodingStandard'

yongchuann committed:
Give users a way to find a task by searching for a keyword

Added FindCommand class
Added searchList function to TaskList
Added new ui functions to Ui

yongchuann committed:
Merge branch 'branch-level-9'

yongchuann committed:
Update the coding style to adhere to the coding style

yongchuann committed:
Merge branch 'branch-A-JavaDoc'

yongchuann committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-gradle

yongchuann committed:
Use Gradle to automate some of the build tasks of the project

yongchuann committed:
Use checkStyle to detect coding style violations

yongchuann committed:
Merge branch 'add-gradle-support'

yongchuann committed:
Add a GUI to Duke

To be updated to remove redundant code and classes
To add a new interface for Duke

yongchuann committed:
Add a GUI to Duke
Add new interface to Pika (previously Duke)
Update code style

yongchuann committed:
Merge branch 'branch-level-10'

yongchuann committed:
Add a GUI to Duke
Add new interface to Pika (previously Duke)
Add user interface as Ash
Fix minor bugs
Change colours and GUI interface

NekrozQliphort commented on src/main/java/pika/ui/TaskList.java:
```diff
@@ -0,0 +1,148 @@
+
+    /**
+     * Searches the list for any task with the given string
+     * @param string Input string pattern to find
```
Should there be an empty line between description and parameter section?

NekrozQliphort commented on src/main/java/pika/ui/Storage.java:
```diff
@@ -0,0 +1,54 @@
+
+    /**
+     * Constructor for Storage Class
+     * @param path the filepath where the txt file will be/is created
```
Perhaps a newline in between the description and parameters? I noticed the same issue in several other places too.

NekrozQliphort replied:
Perhaps a newline in between the description and parameters?

NekrozQliphort commented on src/main/java/pika/ui/TaskList.java:
```diff
@@ -0,0 +1,148 @@
+
+public class TaskList { //TaskList class used to store the tasks and will be updated from the Command class
+    protected int count;
+    private final ArrayList<Task> listArray;
```
Perhaps `listArray` would be better named as `tasks`? In my opinion, this helps clarify it being a multi-valued variable as well.

NekrozQliphort replied:
Perhaps `listArray` would be better named as `tasks`? In my opinion, this helps clarify it being a multi-valued variable as well.

NekrozQliphort commented on src/main/java/pika/Pika.java:
```diff
@@ -0,0 +1,205 @@
+public class Pika extends Application {
+    private final Storage storage;
+    private TaskList taskLists;
+    private final Ui ui;
```
Should `storage` and `ui` perhaps be all uppercase as they are constants?

NekrozQliphort replied:
Should `storage` and `ui` perhaps be all uppercase as they are constants?

NekrozQliphort commented on src/main/java/pika/ui/TaskList.java:
```diff
@@ -0,0 +1,148 @@
+
+    /**
+     * Searches the list for any task with the given string
+     * @param string Input string pattern to find
```
Should there be an empty line between description and parameter section? I noticed the same issue in several other places too.

yongchuann replied:
Yeap! Thanks for pointing that out!

NekrozQliphort commented on src/main/java/pika/ui/TaskList.java:
```diff
@@ -0,0 +1,148 @@
+
+    /**
+     * Returns the list of task for the user
+     */
```
Should there be a @return here to clarify what is it returning?

yongchuann replied:
Yes! thanks for catching that too !!

NekrozQliphort reviewed (commented):
Overall, I found your code follows the guideline given really well, with only a few violations even with quite a lot of code. I only had to point out a few minor coding standard violations, but do give them a look.

djiangel commented on src/main/java/pika/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Marks the task as done when called and return it
+     */
+    public String done() {
```
Naming it markAsDone or completeTask might be more descriptive?

yongchuann replied:
Yes thanks for pointing this out!

djiangel commented on src/main/java/pika/ui/Ui.java:
```diff
@@ -0,0 +1,86 @@
+    public static String showLoadingError() {
+        return "Pika pi!! It seems like your file format is incorrect!";
+    }
+}
```
In general,  I feel that there can be a better naming structure. Some functions have the word show but some doesn't.

yongchuann replied:
Hi thanks for pointing that out! I'll change accordingly!

djiangel reviewed (commented):
LGTM! Only two small points I could point out. (:

xlunarin commented on src/main/java/pika/Pika.java:
```diff
@@ -0,0 +1,205 @@
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.Region;
+import javafx.scene.layout.VBox;
+import javafx.stage.Stage;
```
Maybe you could add an empty line here to separate between `import javafx` and `import pika` inorder to improve readability?

xlunarin commented on src/main/java/pika/command/DeleteCommand.java:
```diff
@@ -0,0 +1,54 @@
+
+
+    /**
+     * Constructor for the DeleteCommand
```
Shouldn't javadoc sentences end with a full stop? 😃

yongchuann replied:
Ah yes thanks as well HAHA

xlunarin commented on src/main/java/pika/command/Command.java:
```diff
@@ -0,0 +1,45 @@
+     *
+     * @param taskList The current list of tasks
+     * @param ui       The current Ui
+     * @param storage  The current storage class to handle the txt file
```
The return variable is missing from the javadoc comment.
```java
* @return ...
```

yongchuann replied:
Thanks for the catch! I forgot to update when i changed all of those to return a string instead of just printing it out

xlunarin reviewed (commented):
Overall nice and neat code!!! I like that you add a lot of comments makes your code easier to understand 😁

yongchuann committed:
Add assertions to taskList, Storage, Ui, TaskTime methods

There are several methods that hold certain assumptions that their input will be of the right format. The assertions are there to ensure that if any of those assumption fails, it will be caught

Let's also check if there are any other functions that needs to have an assertion due to the inputs.

yongchuann committed:
Use refactor to improve the code quality

Update most of the javadocs to adhere to coding standards
Update variable so that those that should not be final are not
Update naming convention for some function to improve clarity

yongchuann committed:
Merge pull request #1 from random689/branch-A-Assertion

Add assertions to taskList, Storage, Ui, TaskTime methods

yongchuann committed:
Merge pull request #2 from random689/branch-A-CodeQuality

Use refactor to improve the code quality

yongchuann committed:
Provide a way to tag task

Add a Tag class
Update Task to store tags and print tags
Update TaskList to account for tags as well when printing out the taskList
Update Storage and TaskList to be able to read tags from previous files
Update variable so that those that should not be final are not
Update naming convention for some function to improve clarity

yongchuann committed:
Merge pull request #3 from random689/branch-C-Tags

Provide a way to tag task

yongchuann committed:
Set theme jekyll-theme-modernist

yongchuann committed:
Add a user guide

Update font size and font

yongchuann committed:
Merge pull request #5 from random689/branch-A-UserGuide

Add a user guide

yongchuann committed:
Improve GUI and bugfix

Add background, error message bubble, and update previous message bubble
Fix several bugs due to the addition of tags
Update JavaDoc for some to adhere to standards

yongchuann committed:
Merge pull request #6 from random689/branch-A-BetterGui

Improve GUI and bugfix

yongchuann committed:
Improve GUI and bugfix

Update JavaDoc for some to adhere to standards
Update MANIFEST.MF main class

yongchuann committed:
Merge pull request #7 from random689/branch-A-Release

Improve GUI and bugfix

yongchuann committed:
Update user guide

yongchuann committed:
Update user guide
Fix a bug with Todo task

Add test files as it was deleted by accident
